README.md
# Aelog
An simple, async, full package name path, log rotating, different colored log library.

aelog aims to make using python log as simple as possible. as a result, it drastically 
simplifies using python logging.

aelog's design objectives:

- Make using python log as simple as possible.
- Output log contains the full package name path.
- Provide asynchronous log output function, at the same time, contains common log output.
- Output according to the log level to mark the different colors separately.
- Provide a log file rotating, automatic backup.
- Output to the terminal and file, default output to the terminal, if you don't provide the log file path.

# Installing aelog
- ```pip install aelog```

# init aelog 
```
import aelog 

app = Flask(__name__)

aelog.init_app(app)
# or 
aelog.init_app(aelog_access_file='aelog_access_file.log', aelog_error_file='aelog_error_file.log', 
               aelog_console=False)
```
# aelog config
List of configuration keys that the aelog extension recognizes:


| configuration key | the meaning of the configuration key |
| ------ | ------ |
| AELOG_ACCESS_FILE | Access file path, default None. |
| AELOG_ERROR_FILE | Error file path, default None. |
| AELOG_CONSOLE | Whether it is output at the terminal, default false. |
| AELOG_LEVEL | log level, default 'DEBUG'. |
| AELOG_MAX_BYTES | Log file size, default 50M. |
| AELOG_BACKUP_COUNT | Rotating file count, default 5.|

# Usage
### simple using, output log to terminal.
```
import aelog

aelog.init_app(aelog_console=True)

def test_aelog_output_console():
    """

    Args:

    Returns:

    """
    aelog.debug("simple debug message", "other message")
    aelog.info("simple info message", "other message")
    aelog.warning("simple warning message", "other message")
    aelog.error("simple error message", "other message")
    aelog.critical("simple critical message", "other message")
    try:
        5 / 0
    except Exception as e:
        aelog.exception(e)
```
This will output to the terminal.  
![console](https://raw.githubusercontent.com/tinybees/aelog/master/docs/output_console.png)
- Different levels of logging, different color, the color is cyan, green, yellow, red and 'bold_red,bg_white' in turn.

### output log to file and terminal.
```
import aelog
from flask import Flask

app = Flask(__name__)

aelog.init_app(app)  # Output to the test.log file and terminal 

def test_aelog_output_file():
    """

    Args:

    Returns:

    """
    aelog.debug("simple debug message", "other message")
    aelog.info("simple info message", "other message")
    aelog.warning("simple warning message", "other message")
    aelog.error("simple error message", "other message")
    aelog.critical("simple critical message", "other message")
    try:
        5 / 0
    except Exception as e:
        aelog.exception(e)
```
This will output to the test.log file and terminal.
![console](https://raw.githubusercontent.com/tinybees/aelog/master/docs/output_file.png)
- Automatic output is greater than the error information to the 'test_error.log' file.
- Different levels of logging, different color, the color is cyan, green, yellow, red and 'bold_red,bg_white' in turn.

### asynchronous output log to file and terminal.
```
import asyncio
import aelog
from sanic import Sanic

app = Sanic(__name__)

aelog.init_aelog(app)  # Output to the test.log file and terminal 

async def test_async_output():
    await aelog.async_debug("simple debug message", "other message")
    await aelog.async_info("simple info message", "other message")
    await aelog.async_warning("simple warning message", "other message")
    await aelog.async_error("simple error message", "other message")
    await aelog.async_critical("simple critical message", "other message")
    try:
        5 / 0
    except Exception as e:
        await aelog.async_exception(e)

if "__name__"=="__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(test_async_output())
```
This will output to the test.log file and terminal.
![console](https://raw.githubusercontent.com/tinybees/aelog/master/docs/async_output.png)
- Automatic output is greater than the error information to the 'test_error.log' file.  
- Different levels of logging, different color, the color is cyan, green, yellow, red and 'bold_red,bg_white' in turn.


CHANGES.md
## Aelog Changelog

###[1.0.9] - 2020-11-12

#### Added
- 增加适配fastapi框架直接从state的config中读取配置的功能

###[1.0.8] - 2020-9-27

#### Added
- 新增sanic日志格式以支持sanic的access日志输出
- 增加类型标注
- 删除废弃的功能
- 增加从app中直接读取log level的功能
- 更改readme文件

###[1.0.6] - 2020-8-30

#### Added
- 新增输出日志的级别开关功能
- 新增console输出也能增加日志级别的功能

#### Changed
- 修复文件中输出的日志无法输出debug日志的情况


###[1.0.5] - 2020-6-27

#### Changed
- 修复如果打印的log参数不是字符串会报错的问题

###[1.0.4] - 2020-6-14

#### Added 
- 增加日志中可以直接输入多个参数的功能,除了exception级别的日志.

#### Changed
- 优化各个级别的日志中影响性能的判断逻辑.
- 删除已经废弃,不再支持的功能.
- 修改最低支持的Python版本为3.6版本.

###[1.0.3] - 2018-12-23

#### Added 

- 增加init_app初始化方法，可以直接传入app初始化（参考类flask扩展的方式）
- 增加指定错误日志文件的功能，如没有指定则和老版本处理方式相同

#### Changed
- 修改init_aelog实现方式，增加警告内容
- 增加log文件的校验，保证是log结尾的日志文件
- 重构部分内容

###[1.0.1] - 2018-03-28

#### Added 

- 测试完成，发布到pypi


#### Changed
- 修改直接在终端执行报错的问题

###[1.0.0] - 2018-03-25

#### Added 

- Make using python log as simple as possible.
- Output log contains the full package name path.
- Provide asynchronous log output function, at the same time, contains common log output.
- Output according to the log level to mark the different colors separately.
- Provide a log rotating, automatic backup.
- Default output to the terminal, if you don't provide the log file path.
- Edit readme,push to pypi, and so on.


#### Changed


tests/tests.py
#!/usr/bin/env python3
# coding=utf-8

"""
@author: guoyanfeng
@software: PyCharm
@time: 18-3-25 下午7:53
"""



tests/verify.py
#!/usr/bin/env python3
# coding=utf-8

"""
@author: guoyanfeng
@software: PyCharm
@time: 18-3-25 下午7:53
"""

import asyncio

from tests import test_aelog

if __name__ == '__main__':
    test_aelog.test_aelog_output_console()
    test_aelog.test_aelog_output_file()
    loop = asyncio.get_event_loop()
    loop.run_until_complete(test_aelog.test_async_output())


tests/__init__.py
#!/usr/bin/env python3
# coding=utf-8

"""
@author: guoyanfeng
@software: PyCharm
@time: 18-3-25 下午7:54
"""



aelog/log.py
#!/usr/bin/env python3
# coding=utf-8

"""
@author: guoyanfeng
@software: PyCharm
@time: 18-3-24 下午10:24
"""
import sys
# noinspection PyProtectedMember
from logging import _nameToLevel
from typing import Dict

from .consts import BACKUP_COUNT, MAX_BYTES

__all__ = ["aelog_config", "aelog_default_config", "sanic_log_config"]


def verify_loglevel(loglevel: str) -> str:
    """
    校验loglevel是否正确
    Args:
        loglevel: log level name
    Returns:

    """
    level_name = list(_nameToLevel.keys())
    loglevel = loglevel.upper()
    if loglevel not in level_name:
        raise ValueError(f"参数loglevel必须为{level_name}中的一个")
    return loglevel


def aelog_default_config(loglevel: str = "DEBUG") -> Dict:
    """
    default logging config
    Args:
        loglevel: log level, default debug
    Returns:

    """
    loglevel = verify_loglevel(loglevel)
    return {
        "version": 1,
        "disable_existing_loggers": False,

        "formatters": {
            "aelog_default": {
                "format": '%(asctime)s %(log_color)s [%(levelname)s] %(name)s [%(funcName)s %(lineno)d]: %(message)s',
                "datefmt": "[%Y-%m-%d %H:%M:%S %z]",
                "class": "colorlog.ColoredFormatter",
                "reset": True,
                "log_colors": {
                    'DEBUG': 'cyan',
                    'INFO': 'green',
                    'WARNING': 'yellow',
                    'ERROR': 'red',
                    'CRITICAL': 'bold_red,bg_white',

                },
            }
        },
        "handlers": {
            "aelog_console": {
                "class": "logging.StreamHandler",
                "formatter": "aelog_default",
                "stream": sys.stdout,
            }
        },
        "loggers": {
            "": {
                "level": loglevel,
                "handlers": ["aelog_console"]
            }
        }
    }


def aelog_config(access_file: str, *, error_file: str = None, console: bool = True, loglevel: str = "DEBUG",
                 max_bytes: int = MAX_BYTES, backup_count: int = BACKUP_COUNT) -> Dict:
    """
    global logging config
    Args:
        access_file: access log full file
        console: terminal output log
        max_bytes: log file max bytes
        backup_count: backup count
        error_file: error log full file
        loglevel: log level, default debug
    Returns:

    """
    loglevel = verify_loglevel(loglevel)
    if console:
        handlers = ["aelog_console", "aelog_access_file", "aelog_error_file"]
    else:
        handlers = ["aelog_access_file", "aelog_error_file"]
    access_file = access_file if access_file.endswith(".log") else "{}.{}".format(access_file, "log")
    if error_file is None:
        pre_file_name, expanded_name = access_file.rsplit(".")
        error_file = "{}.{}".format("{}_error".format(pre_file_name), expanded_name)
    else:
        error_file = error_file if error_file.endswith(".log") else "{}.{}".format(error_file, "log")

    log_config = {
        "version": 1,
        "disable_existing_loggers": False,

        "formatters": {
            "aelog_default": {
                "format": '%(asctime)s %(log_color)s [%(levelname)s] %(name)s [%(funcName)s %(lineno)d]: %(message)s',
                "datefmt": "[%Y-%m-%d %H:%M:%S %z]",
                "class": "colorlog.ColoredFormatter",
                "reset": True,
                "log_colors": {
                    'DEBUG': 'cyan',
                    'INFO': 'green',
                    'WARNING': 'yellow',
                    'ERROR': 'red',
                    'CRITICAL': 'bold_red,bg_white',
                },
            },
        },
        "handlers": {
            "aelog_console": {
                "class": "logging.StreamHandler",
                "formatter": "aelog_default",
                "stream": sys.stdout,
            },
            "aelog_access_file": {
                "class": "logging.handlers.RotatingFileHandler",
                "formatter": "aelog_default",
                "filename": access_file,
                "maxBytes": max_bytes,
                "backupCount": backup_count,
                "encoding": "utf8"
            },
            "aelog_error_file": {
                "class": "logging.handlers.RotatingFileHandler",
                "level": "ERROR",
                "formatter": "aelog_default",
                "filename": error_file,
                "maxBytes": max_bytes,
                "backupCount": backup_count,
                "encoding": "utf8"
            },
        },
        "loggers": {
            "": {
                "level": loglevel,
                "handlers": handlers
            }
        }
    }
    return log_config


def sanic_log_config(access_file: str, *, error_file: str = None, console: bool = True, loglevel: str = "DEBUG",
                     max_bytes: int = MAX_BYTES, backup_count: int = BACKUP_COUNT) -> Dict:
    """
    global logging config
    Args:
        access_file: access log full file
        console: terminal output log
        max_bytes: log file max bytes
        backup_count: backup count
        error_file: error log full file
        loglevel: log level, default debug
    Returns:

    """
    loglevel = verify_loglevel(loglevel)
    if access_file.endswith(".log"):
        access_file = access_file
    else:
        access_file = "{}.{}".format(access_file, "log")

    if error_file is None:
        pre_file_name, expanded_name = access_file.rsplit(".")
        error_file = "{}.{}".format("{}_error".format(pre_file_name), expanded_name)
    else:
        error_file = error_file if error_file.endswith(".log") else "{}.{}".format(error_file, "log")

    log_config = {
        "version": 1,
        "disable_existing_loggers": False,

        "formatters": {
            "generic": {
                "format": "%(asctime)s [%(process)d] [%(levelname)s] %(message)s",
                "datefmt": "[%Y-%m-%d %H:%M:%S %z]",
                "class": "logging.Formatter",
            },
            "access": {
                "format": "%(asctime)s - (%(name)s)[%(levelname)s][%(host)s]: "
                          + "%(request)s %(message)s %(status)d %(byte)d",
                "datefmt": "[%Y-%m-%d %H:%M:%S %z]",
                "class": "logging.Formatter",
            },
        },
        "handlers": {
            "console": {
                "class": "logging.StreamHandler",
                "formatter": "generic",
                "stream": sys.stdout,
            },
            "access_console": {
                "class": "logging.StreamHandler",
                "formatter": "access",
                "stream": sys.stdout,
            },
            "access_file": {
                "class": "logging.handlers.RotatingFileHandler",
                "level": "INFO",
                "formatter": "access",
                "filename": access_file,
                "maxBytes": max_bytes,
                "backupCount": backup_count,
                "encoding": "utf8"
            },
            "error_console": {
                "class": "logging.StreamHandler",
                "formatter": "generic",
                "stream": sys.stdout,
            },
            "error_file": {
                "class": "logging.handlers.RotatingFileHandler",
                "level": "ERROR",
                "formatter": "generic",
                "filename": error_file,
                "maxBytes": max_bytes,
                "backupCount": backup_count,
                "encoding": "utf8"
            },
        },
        "loggers": {
            "sanic.root": {
                "level": loglevel,
                "handlers": ["console"] if console else []},
            "sanic.error": {
                "level": loglevel,
                "handlers": ["error_file", "error_console"] if console else ["error_file"],
                "qualname": "sanic.error",
            },
            "sanic.access": {
                "level": "INFO",
                "handlers": ["access_file", "access_console"] if console else ["access_file"],
                "qualname": "sanic.access",
            },
        }
    }

    return log_config


aelog/aelog.py
#!/usr/bin/env python3
# coding=utf-8

"""
@author: guoyanfeng
@software: PyCharm
@time: 18-3-24 下午10:23
"""

import asyncio
import inspect
import logging
import os
import traceback
from concurrent.futures import ThreadPoolExecutor
from functools import partial
from logging.config import dictConfig
from typing import Dict, Tuple

from .consts import BACKUP_COUNT, MAX_BYTES
from .log import aelog_config, aelog_default_config

__all__ = ("init_app", "debug", "info", "warning", "error", "critical", "exception", "async_debug",
           "async_info", "async_warning", "async_error", "async_exception", "async_critical")

_pool = ThreadPoolExecutor()


def init_app(app=None, *, aelog_access_file: str = None, aelog_error_file: str = None,
             aelog_console: bool = True, aelog_level: str = "DEBUG",
             aelog_max_bytes: int = MAX_BYTES, aelog_backup_count: int = BACKUP_COUNT):
    """
    init global logging

    if aelog_access_file is none, then output log to the terminal.

    Args:
        app: app 应用
        aelog_access_file: access log full file
        aelog_error_file: error log full file
        aelog_level: log level, default debug
        aelog_console: terminal output log
        aelog_max_bytes: log file max bytes
        aelog_backup_count: backup count
    Returns:

    """
    if app is not None:
        config: Dict = app.config if getattr(app, "config", None) else app.state.config
        aelog_access_file = config.get("AELOG_ACCESS_FILE") or aelog_access_file
        aelog_error_file = config.get("AELOG_ERROR_FILE") or aelog_error_file
        aelog_console = config.get("AELOG_CONSOLE", aelog_console)
        aelog_level = config.get("AELOG_LEVEL") or aelog_level
        aelog_max_bytes = config.get("AELOG_MAX_BYTES") or aelog_max_bytes
        aelog_backup_count = config.get("AELOG_BACKUP_COUNT") or aelog_backup_count

    if aelog_access_file is None:
        aelog_conf = aelog_default_config(loglevel=aelog_level)
    else:
        aelog_conf = aelog_config(aelog_access_file, error_file=aelog_error_file, console=aelog_console,
                                  loglevel=aelog_level, max_bytes=aelog_max_bytes, backup_count=aelog_backup_count)
    dictConfig(aelog_conf)


def find_caller(caller_frame, stack_info=False) -> Tuple:
    """
    Find the stack frame of the caller so that we can note the source
    file name, line number and function name.
    """
    if hasattr(caller_frame, "f_code"):
        co = caller_frame.f_code
    else:
        return "(unknown file)", 0, "(unknown function)", None

    sinfo = None
    if stack_info:
        sio = os.io.StringIO()  # type: ignore
        sio.write('Stack (most recent call last):\n')
        traceback.print_stack(caller_frame, file=sio)
        sinfo = sio.getvalue()
        if sinfo[-1] == '\n':
            sinfo = sinfo[:-1]
        sio.close()
    return co.co_filename, caller_frame.f_lineno, co.co_name, sinfo


def debug(msg, *args, sep=' '):
    """
    Log a message with severity 'DEBUG' on the root logger. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        sep: 多个消息的分隔符, 默认为一个空格.
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    logger.debug(f"{sep}".join(str(val) for val in (msg, *args)))


def info(msg, *args, sep=' '):
    """
    Log a message with severity 'INFO' on the root logger. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        sep: 多个消息的分隔符, 默认为一个空格.
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    logger.info(f"{sep}".join(str(val) for val in (msg, *args)))


def warning(msg, *args, sep=' '):
    """
    Log a message with severity 'WARNING' on the root logger. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        sep: 多个消息的分隔符, 默认为一个空格.
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    logger.warning(f"{sep}".join(str(val) for val in (msg, *args)))


def error(msg, *args, sep=' '):
    """
    Log a message with severity 'ERROR' on the root logger. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        sep: 多个消息的分隔符, 默认为一个空格.
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    logger.error(f"{sep}".join(str(val) for val in (msg, *args)))


def critical(msg, *args, sep=' '):
    """
    Log a message with severity 'CRITICAL' on the root logger. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        sep: 多个消息的分隔符, 默认为一个空格.
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    logger.critical(f"{sep}".join(str(val) for val in (msg, *args)))


def exception(msg, *args, **kwargs):
    """
    Log a message with severity 'ERROR' on the root logger, with exception information. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        kwargs: 其他参数
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    logger.exception(msg, *args, exc_info=msg, **kwargs)


async def async_debug(msg, *args, sep=' '):
    """
    Log a message with severity 'DEBUG' on the root logger. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        sep: 多个消息的分隔符, 默认为一个空格.
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    await asyncio.wrap_future(_pool.submit(logger.debug, f"{sep}".join(str(val) for val in (msg, *args))))


async def async_info(msg, *args, sep=' '):
    """
    Log a message with severity 'INFO' on the root logger. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        sep: 多个消息的分隔符, 默认为一个空格.
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    await asyncio.wrap_future(_pool.submit(logger.info, f"{sep}".join(str(val) for val in (msg, *args))))


async def async_warning(msg, *args, sep=' '):
    """
    Log a message with severity 'WARNING' on the root logger. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        sep: 多个消息的分隔符, 默认为一个空格.
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    await asyncio.wrap_future(_pool.submit(logger.warning, f"{sep}".join(str(val) for val in (msg, *args))))


async def async_error(msg, *args, sep=' '):
    """
    Log a message with severity 'ERROR' on the root logger. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        sep: 多个消息的分隔符, 默认为一个空格.
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    await asyncio.wrap_future(_pool.submit(logger.error, f"{sep}".join(str(val) for val in (msg, *args))))


async def async_critical(msg, *args, sep=' '):
    """
    Log a message with severity 'CRITICAL' on the root logger. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        sep: 多个消息的分隔符, 默认为一个空格.
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    await asyncio.wrap_future(_pool.submit(logger.critical, f"{sep}".join(str(val) for val in (msg, *args))))


async def async_exception(msg, *args, **kwargs):
    """
    Log a message with severity 'ERROR' on the root logger, with exception information. If the aelog has no initialize,
    call init_app() and output to the terminal.
    Args:
        msg: 要打印消息内容
        args: 要打印的其他消息内容
        kwargs: 其他参数
    Returns:

    """
    caller_frame = inspect.currentframe().f_back
    caller_module = inspect.getmodule(caller_frame)
    logger = logging.getLogger(caller_module.__name__ if caller_module else "")
    logger.findCaller = partial(find_caller, caller_frame)
    await asyncio.wrap_future(_pool.submit(logger.exception, msg, *args, exc_info=msg, **kwargs))
